### DISCUSSION TITLE
Brush discussion

### DISCUSSION THREAD

**Discussion by Nukiloco** [opentoonz/opentoonz#6747] [Brushes]
In my opinion out of any raster program I have used, how the brushes work and are handled in this program in my opinion is really not that great right now. I have laid out a few complaints of what I think about the brush syetem. In some of them I laid out some possible solutions that I think would work out.

First complaint:
WO Brush (default brush) doesn't match up with any of the other custom brushes. Which is just weird in my opinion, if the majority or almost every single brush is going to use the same system, then the WO Brush should use it too or it shouldn't even be here at all in the first place. I assume the brush was only made for Toonz Raster or Toonz Vector?

I really think there needs to be a discussion about this as this brush can confuse new beginners like me to this program that it just works differently than all the other brushes? In my opinion that doesn't really sound right at all and more in fact confuses a lot of beginners who use the program.

Second complaint:
Top simple brush settings? Aka where the Size, Opacity, Hardness, and sometimes Smoothness option is located at the top at. What's weird is that when you select any brush than the WO Brush, is that first the Size property is at 0? I feel like the size property should atleast match up with the current size of the brush instead of what I assume to be it just multiplying? some values in the brush settings. The brush highlight seems to be able to detect the current brush size so I assume the program can detect it also unless if I'm wrong about that.

Third complaint:
It seems like depending on the type of raster level I'm on makes a difference in how the brushes work, aka Toonz Raster doesn't work with any of the other custom brushes while Raster works with most custom brushes? I'm confused on why that would happen like that. I'm assuming both of them have advantages and disadvantages in how they work? I now tested what a Toonz Raster does and it seems like when changing a color changes all the colors that you drew with on that frame. While on a regular Raster level, it wouldn't swap out all the colors which just works like any other raster program. I'm not sure what advantages that swapping out the color can give, the only advantages I can think of where that you can swap out colors in animation sketches though I still don't see the use in doing that? If people wanted to swap out colors, there should be layer or a level option that allows them to swap a certain color out with another color. It doesn't seem like there is really any other point in using a Toonz Raster level?

Fourth complaint:
Coming back to the brush settings again and discovering more about it, I think each brush in a Raster level should have the simple settings of having a Size (actually matching up with the brush size instead of being 0), Smoothness, Hardness, and Opacity instead of some of options like Smoothness being hidden in the advanced brush settings. I figured out that one of the properties aka Slow position tracking, is the most effective right now at simulating a smoothness option. I think the Smoothness property should match up with how Slow position tracking works, or it should mimic with how the WO Brush works. There is another option called Slow tracking per dab but it doesn't seem to affect anything with any brushes at all from what I tested with it.

That's the current list of complaints that I have right now with the brush system in OpenToonz. Feel free to discuss!

**Comment by beeheemooth** [comment]
What is "WO Brush"?

**Comment by artisteacher** [comment]
The default brush, I think?

1st complaint:
Originally, the default raster brush was the only brush available for raster levels. MyPaint brushes were only added fairly recently. Long-time users didn't want to lose certain behaviors they relied on, so the "default brush" was left as is, even though it behaves differently. I assume the MyPaint brushes were programmed according to how they are used in the MyPaint program, so that new brushes can be ported over easily without needing much adjustment. The default brush isn't going away and isn't especially likely to change. Making significant changes to the behavior of the MyPaint brushes might actually be limiting...


3rd complaint:
Toonz raster levels are actually a little more complex. The color mapping itself is helpful if you are reusing certain assets for something new and want to make changes. The entire palette can actually be switched out in one go. Additionally, the color mapping actually allows you to easily keyframe color transitions and see them live. Due to the mapping, you can also use texture styles.

Not only do the toonz raster levels keep track of colors, they also keep track of fills vs. strokes. This is great for cel shading techniques - especially in combination with the paint brush tool. You can fill or paint on just the stroke, just the fill, or both.

Due to the color/stroke/fill mapping, brushes for blending, smudging and erasing don't currently work on the toonz raster levels. That's where the standard raster levels come in - they are best for a traditional or painterly workflow. However, colors/strokes/fills are not mapped. So there's no animating colors directly and painted colors can't be changed just by changing the style.

**Comment by RodneyBaker** [comment]
You may be interested in developments over in the Tahoma2D fork relating to brushes here: https://github.com/tahoma2d/tahoma2d/pull/1915

That is available in the T2D nightly and after the official release (est. December) might be a good candidate for porting to Opentoonz.
